File: game/piece.go
package game

import "fmt"

type PieceType struct {
	Name string
}

func (pt *PieceType) String() string {
	return pt.Name
}

type Piece struct {
	Type  *PieceType
	Owner *Player
}

func (p *Piece) String() string {
	return fmt.Sprintf("%s %s", p.Owner.Color, p.Type)
}

File: fe/src/api/schema/piece.ts
import { Piece } from "@/model/game/piece";
import { ColorDto, colorToModel } from "./color";
import { PieceTypeDto, pieceTypeToModel } from "./pieceType";
import { SquareDto, squareToModel } from "./square";

export interface PieceDto {
  Type: PieceTypeDto;
  Color: ColorDto;
  Square: SquareDto;
}

export const pieceToModel = (piece: PieceDto): Piece => ({
  type: pieceTypeToModel(piece.Type),
  square: squareToModel(piece.Square),
  color: colorToModel(piece.Color),
});
